### ISSUE TITLE
AVPro 3.1.0 requires Android API level 34 which is not supported by Meta Quest

### ISSUE THREAD

**Issue by alex8b** [RenderHeads/UnityPlugin-AVProVideo#2042]
AVPro 3.1.0 is compiled against android-34. 
Unfortunately Meta Quest Horizon OS only supports android-32.
I also can't find a way to download an older version (3.0.x) from the asset store.
Is there a way to download 3.0.x?
Could you downgrade the API level?
Is there maybe some compile switch?

```
> A failure occurred while executing com.android.build.gradle.internal.tasks.CheckAarMetadataWorkAction
   > 20 issues were found when checking AAR metadata:
     
       1.  Dependency 'androidx.media3:media3-exoplayer-dash:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       2.  Dependency 'androidx.media3:media3-exoplayer-hls:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       3.  Dependency 'androidx.media3:media3-exoplayer-rtsp:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       4.  Dependency 'androidx.media3:media3-exoplayer-smoothstreaming:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       5.  Dependency 'androidx.media3:media3-exoplayer-workmanager:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       6.  Dependency 'androidx.media3:media3-exoplayer:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       7.  Dependency 'androidx.media3:media3-extractor:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       8.  Dependency 'androidx.media3:media3-container:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       9.  Dependency 'androidx.media3:media3-datasource-cronet:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
      10.  Dependency 'androidx.media3:media3-datasource-okhttp:1.4.1' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :launcher is currently compiled against android-32.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.3.1 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) a<message truncated>
```

**Comment by Chris-RH** [comment]
Make sure android-32 and android-34 are found in \Editor\Data\PlaybackEngines\AndroidPlayer\SDK\Platforms

In your gradles (main and launcher Templates), write this:

minSdkVersion 22 (or higher if you wish)
targetSDKversion 32
compileSDKversion 34

**Comment by FG148** [comment]
Thanks for youre suggestion. You add this directly in the android section?
Right now I am working with Unity 2022.3.14. would Unity 6 be supported?
```

android {
    ndkPath "**NDKPATH**"
    
    minSdkVersion 22
    targetSDKversion 32
    compileSDKversion 33
    buildToolsVersion '**BUILDTOOLS**'
```

Somehow after changing the gradles always get error messages:

___FAILURE: Build completed with 2 failures.
1: Task failed with an exception.

    Where:
    Build file '...\Library\Bee\Android\Prj\IL2CPP\Gradle\launcher\build.gradle' line: 7

    What went wrong:
    A problem occurred evaluating project ':launcher'.

    No signature of method: build_euip490diy9gnuhjp550vmghz.android() is applicable for argument types: (build_euip490diy9gnuhjp550vmghz$_run_closure2) values: [build_euip490diy9gnuhjp550vmghz$run_closure2@68dd9d28]
    problem occurred configuring project ':launcher'.
    compileSdkVersion is not specified. Please add it to build.gradle__

I would be grateful for any advice!

**Comment by Ste-RH** [comment]
I have downloaded Unity 2022.3.14 and the following steps enabled me to build a project using the trial version of AVPro Video v3.1.0 ...

1) Open 'Player Settings'
2) Navigate to 'Target API Level' and set it to 'API Level 34'
3) Build
4) You will be prompted that the 'Android APK is missing', Select 'Update Android SDK'.  This will open up a command window and download the required APK version.

![image](https://github.com/user-attachments/assets/0ded68e3-197e-40d7-804f-9efa3fe0550f)

![image](https://github.com/user-attachments/assets/a0d9a882-0031-4c20-a738-209926df5eff)

**Comment by FG148** [comment]
Thank you for your suggestion! The installation of SDK level 34 is indeed necessary for AVProVideo after v.3.

The problem is that for uploading apps for Meta Quest, the target SDK level needs to be set to 32.

I have just tried a few attempts for the mentioned problems and finally came to a solution. I don't understand the logic 
behind it, but it works :-)

Just follow the steps as in [RenderHeads Android platform ](https://www.renderheads.com/content/docs/AVProVideo/articles/platform-android.html#unity-20221-and-above) to create the Custom Main Gradle Template and Custom Launcher Gradle Template. Find them inside the project under assets\Plugins\Android and modify them. 

The solution of Chris-RH didn't work in my case. Everytime I tried to define targetSDKversion 32 in the android section I got errors. Putting it inside the defaultConfig section worked for some reason. Right now I get no compilation errors and the upload could be initiated. 



```
android {
    ndkPath "**NDKPATH**"


    compileSdkVersion 34
    buildToolsVersion '**BUILDTOOLS**'

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_11
        targetCompatibility JavaVersion.VERSION_11
    }

    defaultConfig {
        minSdkVersion **MINSDKVERSION**
        targetSdkVersion 32
        ndk {
            abiFilters **ABIFILTERS**
        }
        versionCode **VERSIONCODE**
        versionName '**VERSIONNAME**'
        consumerProguardFiles 'proguard-unity.txt'**USER_PROGUARD**
    }
```
